### DISCUSSION TITLE
unexpected behavior on select in column_property using correlate_except() and scalar_subquery() with polymorphic classes and subclasses

### DISCUSSION THREAD

**Discussion by Hargrav3s** [sqlalchemy/sqlalchemy#8167] [Usage Questions]
### Describe the bug

when using a scalar_subquery in a column_property that selects from a polymorphic class, specifically in the following case of using a func.count.

### To Reproduce

```python
from sqlalchemy import Column, Integer, String, ForeignKey, select, func
from sqlalchemy.orm import declarative_base, relationship, column_property


Base = declarative_base()

class Customer(Base):
    __tablename__ = 'customer'
    id = Column(Integer, primary_key=True)
    type = Column(String(20))

    __mapper_args__ = {
        'polymorphic_on':'type',
        'polymorphic_identity':'customer'
    }

class Store(Customer):
    __tablename__ = "store"
    id = Column(Integer, ForeignKey('customer.id'), primary_key=True)
    retailer_id = Column(Integer, ForeignKey('retailer.id'))
    retailer = relationship(
        "Retailer",
        back_populates="stores",
        lazy="noload",
        foreign_keys=[retailer_id]
    )

    __mapper_args__ = {
        'polymorphic_identity':'store'
    }


class Retailer(Customer):
    __tablename__ = "retailer"
    id = Column(Integer, ForeignKey('customer.id'), primary_key=True)
    stores = relationship(
        "Store",
        back_populates="retailer",
        foreign_keys=[Store.retailer_id],
        lazy="noload"
    )
    store_count = column_property(
        select([func.count(Store.id)]).
        where(Store.retailer_id==id).
        correlate_except(Store).
        scalar_subquery()
    )

    __mapper_args__ = {
        'polymorphic_identity':'retailer'
    }

print(Retailer.store_count.expression)
```

### Error

Actual:
```
(SELECT count(store.id) AS count_1 
FROM retailer, customer JOIN store ON customer.id = store.id 
WHERE store.retailer_id = retailer.id)
``` 

Expected:
```
(SELECT count(store.id) AS count_1
FROM store 
WHERE store.retailer_id = retailer.id)
```

### Versions

- OS: macOS Monterey 12.4
- Python: 3.9.10
- SQLAlchemy: 1.4.37
- Database: N/A
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): N/A

### Additional context

Discovered this by trying to implement a relationship count on a polymorphic subclass itself as a column property.

**Comment by zzzeek** [comment]
your subquery refers to the column "Retailer.id", therefore when you print the subquery as a standalone SELECT statement, "retailer" must be added to the FROM clause.  

I think you are expecting "correlate" to mean, "never render this FROM clause" but that's not what it means. It means to *correlate* it to an enclosing SELECT statement.   if we render your query in context, "retailer" is correlated:

```py
print(select(Retailer.store_count).select_from(Retailer))
```

```sql
SELECT (SELECT count(store.id) AS count_1 
FROM customer JOIN store ON customer.id = store.id 
WHERE store.retailer_id = retailer.id) AS anon_1 
FROM customer JOIN retailer ON customer.id = retailer.id
```

now there is one less-nice thing about this which is that if you say `select(Retailer.store_count)` all by itself without using `select_from()`, it doesn't know that you want `Retailer` as the outermost FROM clause.  but that's a separate issue.

**Reply by ghost** [reply]
@zzzeek 
Ah, I see what you mean, but how could I get the default behavior of `select(Retailer.store_count)` as the outermost FROM clause in this case?

**Reply by ghost** [reply]
In my case, our main goal is to do `db.session.query(Retailer).all()` and have an accurate store_count on all retailer objects. Was following the example here https://docs.sqlalchemy.org/en/14/orm/mapped_sql_expr.html#using-column-property, had this work in an older version of SQLAlchemy, I believe 1.3, but upgrading causes the polymorphic classes to behave differently in this case

**Reply by zzzeek** [reply]
OK things are working more poorly when you do that, because it's getting confused about the Retailer entity and the Store entity and how correlate is supposed to work there.   not ideal behavior.   also, there seems to be some aggressive ORM logic that happens here in the column_property() if I try to select instead from `Store.__table__` to avoid all that it still replaces Store.id with "company.id".

Anyway, the best way to do this is always to alias what's inside the column_property so that SQLAlchemy has no chance to mix up Retailer with Store.

a good way would be like this:

```py
sa = Store.__table__.alias()

class Retailer(Customer):
    __tablename__ = "retailer"
    id = Column(Integer, ForeignKey('customer.id'), primary_key=True)
    stores = relationship(
        "Store",
        back_populates="retailer",
        foreign_keys=[Store.retailer_id],
        lazy="noload"
    )
    store_count = column_property(
        select(func.count(sa.c.id)).
        where(sa.c.id==id).
        correlate_except(sa).
        scalar_subquery()
    )

    __mapper_args__ = {
        'polymorphic_identity':'retailer'
    }
```

you can also use aliased(Store) but that will try to configure the mappers which you can't do until Retailer is defined, so that way is more verbose like this:

```py
class Retailer(Customer):
    __tablename__ = "retailer"
    id = Column(Integer, ForeignKey('customer.id'), primary_key=True)
    stores = relationship(
        "Store",
        back_populates="retailer",
        foreign_keys=[Store.retailer_id],
        lazy="noload"
    )

    __mapper_args__ = {
        'polymorphic_identity':'retailer'
    }

sa = aliased(Store, flat=True)
Retailer.store_count = column_property(
        select(func.count(sa.id)).
        where(sa.retailer_id==Retailer.id).
        correlate_except(sa).
        scalar_subquery()
    )
```

in either case you get a query like:

```sql
SELECT retailer.id, customer.id AS id_1, customer.type, (SELECT count(store_1.id) AS count_1 
FROM store AS store_1 
WHERE store_1.retailer_id = retailer.id) AS anon_1 
FROM customer JOIN retailer ON customer.id = retailer.id
```

**Reply by zzzeek** [reply]
ive added #8168 to try to accommodate the non-intuitive behavior when using `Store.__table__` directly.

**Reply by Hargrav3s** [reply]
Awesome, thank you so much!
